lib/chat/db/scope/key_scope.ex

```diff
@@ -0,0 +1,184 @@
+        MapSet.member?(hashes, one) or MapSet.member?(hashes, other)
+      end)
+      |> MapSet.new()
+
+    dialog_binhashes =
+      dialog_keys
+      |> Enum.map(fn {:dialogs, hash} ->
+        hash
+        |> Utils.hash()
+        |> Utils.binhash()
+      end)
+      |> MapSet.new()
+
+    snap
+    |> db_keys_stream({:dialog_message, 0, 0, 0}, {:"dialog_message\0", 0, 0, 0})
+    |> Stream.filter(fn {:dialog_message, binhash, _, _} ->
+      MapSet.member?(dialog_binhashes, binhash)
+    end)
+    |> union_set(dialog_keys)
+    |> union_set(acc_set)
+  end
+
+  defp add_rooms(acc_set, snap, {binhashes, hashes}) do
+    with_rooms =
+      snap
+      |> db_keys_stream({:rooms, 0}, {:"rooms\0", 0})
+      |> Stream.filter(fn {:rooms, hash} ->
+        MapSet.member?(hashes, hash)
+      end)
+      |> union_set(acc_set)
+
+    snap
+    |> db_keys_stream({:room_message, 0, 0, 0}, {:"room_message\0", 0, 0, 0})
+    |> Stream.filter(fn {:room_message, room_binhash, _, _} ->
+      MapSet.member?(binhashes, room_binhash)
+    end)
+    |> union_set(with_rooms)
+  end
+
+  defp add_content(acc_set, snap, {_, hashes}) do
+    file_index =
+      snap
+      |> db_keys_stream({:file_index, 0, 0, 0}, {:"file_index\0", 0, 0, 0})
+      |> Stream.filter(fn {:file_index, reader_hash, _file_key, _message_id} ->
+        MapSet.member?(hashes, reader_hash)
+      end)
+      |> MapSet.new()
+
+    file_keys =
+      file_index
+      |> Enum.map(fn {:file_index, _reader_hash, file_key, _message_id} -> file_key end)
+      |> MapSet.new()
+
+    file_chunks =
+      snap
+      |> db_keys_stream({:file_chunk, 0, 0, 0}, {:"file_chunk\0", 0, 0, 0})
+      |> Stream.filter(fn {:file_chunk, file_key, _chunk_start, _chunk_end} ->
+        MapSet.member?(file_keys, file_key)
+      end)
+      |> MapSet.new()
+
+    files =
+      snap
+      |> db_keys_stream({:file, 0}, {:"file\0", 0})
+      |> Stream.filter(fn {:file, file_key} ->
+        MapSet.member?(file_keys, file_key)
+      end)
+      |> MapSet.new()
+
+    chunk_keys =
+      snap
+      |> db_keys_stream(
+        {:chunk_key, {:file_chunk, 0, 0, 0}},
+        {:chunk_key, {:"file_chunk\0", 0, 0, 0}}
+      )
+      |> Stream.filter(fn {:chunk_key, {:file_chunk, file_key, _chunk_start, _chunk_end}} ->
+        MapSet.member?(file_keys, file_key)
+      end)
+      |> MapSet.new()
+
+    memo_index =
+      snap
+      |> db_keys_stream({:memo_index, 0, 0}, {:"memo_index\0", 0, 0})
+      |> Stream.filter(fn {:memo_index, reader_hash, _memo_key} ->
+        MapSet.member?(hashes, reader_hash)
+      end)
+      |> MapSet.new()
+
+    memo_keys =
+      memo_index
+      |> Enum.map(fn {:memo_index, _reader_hash, key} -> key end)
+      |> MapSet.new()
+
+    memos =
+      snap
+      |> db_keys_stream({:memo, 0}, {:"memo\0", 0})
+      |> Stream.filter(fn {:memo, key} ->
+        MapSet.member?(memo_keys, key)
+      end)
+      |> MapSet.new()
```

Comment:
This looks like a pattern shared with `room_invites` at least. May we segregate it in a function?

Reply:
Sure.